src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/AddVenueController.cs

```diff
@@ -0,0 +1,193 @@
+
+        private readonly IMediator _mediator;
+        private readonly JourneyInstanceProvider _journeyInstanceProvider;
+        private readonly IProviderContextProvider _providerContextProvider;
+        private JourneyInstance<AddVenueJourneyModel> _journeyInstance;
+
+        public AddVenueController(
+            IMediator mediator,
+            JourneyInstanceProvider journeyInstanceProvider,
+            IProviderContextProvider providerContextProvider)
+        {
+            _mediator = mediator;
+            _journeyInstanceProvider = journeyInstanceProvider;
+            _providerContextProvider = providerContextProvider;
+            _journeyInstance = _journeyInstanceProvider.GetInstance<AddVenueJourneyModel>();
+        }
+
+        [HttpGet("")]
+        public IActionResult Index([FromQuery] string postcode, [FromQuery] string returnUrl)
+        {
+            _journeyInstance = _journeyInstanceProvider.GetOrCreateInstance(
+                () => new AddVenueJourneyModel(),
+                new PropertiesBuilder()
+                    .Add(ReturnUrlJourneyInstancePropertyKey, returnUrl)
+                    .Build());
+
+            if (!_journeyInstanceProvider.IsCurrentInstance(_journeyInstance))
+            {
+                return RedirectToAction()
+                    .WithProviderContext(_providerContextProvider.GetProviderContext())
+                    .WithJourneyInstanceUniqueKey(_journeyInstance);
+            }
+
+            return View(
+                nameof(PostcodeSearch),
+                new PostcodeSearch.SearchCommand()
+                {
+                    Postcode = postcode
+                });
+        }
+
+        [HttpPost("cancel")]
+        [RequireJourneyInstance]
+        public IActionResult Cancel()
+        {
+            _journeyInstance.Delete();
+
+            if (_journeyInstance.Properties.TryGetValue(ReturnUrlJourneyInstancePropertyKey, out var returnUrlObj) &&
+                returnUrlObj is string returnUrl &&
+                Url.IsLocalUrl(returnUrl))
+            {
+                return Redirect(returnUrl);
+            }
+
+            return RedirectToAction(nameof(ViewVenues.ViewVenuesController.ViewVenues), "ViewVenues")
+                .WithProviderContext(_providerContextProvider.GetProviderContext());
+        }
+
+        [HttpGet("postcode-search")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> PostcodeSearch(PostcodeSearch.SearchCommand command) =>
+            await _mediator.SendAndMapResponse(
+                command,
+                result => result.Match(
+                    errors => this.ViewFromErrors(errors),
+                    vm => View("PostcodeSearchResults", vm)));
+
+        [HttpPost("select-postcode")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> SelectPostcode(PostcodeSearch.SelectCommand command, [FromQuery] string postcode)
```

Comment:
Why is the `postcode` param not included in the command?

Reply:
One is bound from the body and the other from the query param. I can't remember if there was a reason why the latter is from the query param, other than consistency with the earlier pages. I'll look at if it's easy to change.

Reply:
This parameter is gone now with the latest changes.